### ISSUE TITLE
Headphone output has a crackling noise

### ISSUE THREAD

**Issue by Holzhaus** [Holzhaus/mixxx-gh-issue-migration#454]
I am using a desktop an on-board sound card and a creative usb sound card. Headphones are connected to the usb sound card. I get a crackling noise through the headphones. Speakers connected to the on-board sound are fine. I did the following checks. 

In preferences, I changed Master to usb sound card and Headphones to onboard sound card. The noise then came from speakers (Speakers). It appears that whichever device is selected, the noise comes from the device that is selected as headphones. 

I downloaded another program and tested. It works properly. 

I also used two mp3 players in parallel using the two devices and sound came correctly.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2013-07-20T06:59:18Z
 * owner: niranjan-x
 * assignee: daschuer
 * the launchpad url was https://bugs.launchpad.net/bugs/1203249

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
This is a known artifact of using two separate audio interfaces. Each one has its own clock crystal and no two are precisely the same frequency even if the devices are the same model and from the same production run. Mixxx synchronizes its audio generation to the clock crystal of whichever device is selected as the master output (deck 1 output if no master is selected) so that the crowd won't hear the artifacts. So the secondary device will either fall behind or run ahead of the primary one, causing it to play silence until Mixxx generates the next audio buffer exactly in time for the primary device. Playing bits of audio interspersed with bits of silence sounds like crackling.
The only good way to resolve this is to use a single audio device with multiple outputs (as then they all run at the same speed since they use a single clock crystal.)

**Comment by Holzhaus** [comment]
(by daschuer)
Thank you Sean for explanations. I am suffering the same issue, when I try to use the internal sound card + the  RMX2 sound card.
I think won't fix or triaged is a better status for it, because it is actual an issue.

Is there a chance to fix it for the preview player, since it plays exclusively to headphones? 
Do you have an idea how to make the effect less noticeable? 
 
     
@Niranjan: Did you play the same source to both audio interfaces, with the other program? Which program do you take for test?

**Comment by Holzhaus** [comment]
(by niranjan-x)
Thanks for the explanation. But it does not happen with virtual DJ home edition which is free. So I was thinking it may be worthwhile pointing this out. 

@Daniel: Yes I did play the same source in both and actually tried a few different files. I used virtual DJ home. This gives you 10 minutes to use headphones and main. After that it switches back to a single source. I guess its because its home edition and Free! However, during the 10 minutes on several clips I didn't notice it.

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
Wow, that _is_ interesting that VDJ has been able to overcome the problem. I actually have thought of a way but it involves re-working how the Mixxx engine operates (so that it generates streams of audio not locked to any sound card.) See the timer-powered engine blueprint I filed awhile ago for more details on that approach.

**Comment by Holzhaus** [comment]
(by daschuer)
I think there is no need for a independent timer, It is just fine to be bound on the master sound interface. 
But there schould be an other way than skip a complete buffer  to adapt to the other interface. Maybe we can use the linear scaler?

An other related bugs are Bug #198306 and probably Bug #194021.

**Comment by Holzhaus** [comment]
(by ywwg)
If we can't figure out a way around the issue, we should create some UI information to tell people about this problem.  If they select a different sound card for the headphones, we should draw a warning icon next to the selection and print some information at the bottom of the window explaining the problem or perhaps providing a link.

the linear scalar relies on the caching reader right now so it would be tough to make that stretch an arbitrary buffer.  However a linear stretcher is a basic algorithm so it shouldn't be hard to write a sampleutil version of a linear scalar.

**Comment by Holzhaus** [comment]
(by daschuer)
Analyse: 

Currently, using two sound cards, we have two concurrent real-time threads. 
Only the thread of the device == m_pClkRefDevice drives the engine an populates the Buffers with new data.

I am not sure, but I am afraid the other thread goes wild and takes the buffer as it is in the specific moments.  
So It copies either the correct buffer, an outdated buffer or something in between to the sound card. 
So it looks like that the Cristal interference make things worse, but mainly we have a simple race condition. 

In most cases, the threads will honest alternate, but if they are scheduled on different CPUs or the timing difference swap the cycle this produces unwanted sounds. 
  
Possible solution: 
Lets only use one Real-Time Callback for the sound card driving master channel. Feed the other streams by the Read/Write I/O Method.   

http://portaudio.com/docs/v19-doxydocs/api_overview.html

It should be enough, to insert or remove samples to sync both streams, because the timing difference might be not that big. A lousy 200 ppm crystal will not produce more than 1 sample distorsion per callback  at 100 ms audio buffer size.       

What do you think?

**Comment by Holzhaus** [comment]
(by daschuer)
It seems Bug #900364 suffers the same issue.

**Comment by Holzhaus** [comment]
(by daschuer)
Bug #726894 is related

**Comment by Holzhaus** [comment]
(by mike-globalfunkradio)
Only fix is to buy a single clock dual Soundcard, or use Virtual DJ, last post on this bug shows here as 2013-11-11, is there any more information on resolution yet? or another thread I can follow?

**Comment by Holzhaus** [comment]
(by daschuer)
No, Sorry right thread but no update yet. But it is scheduled for the upcoming Mixxx version.

**Comment by Holzhaus** [comment]
(by mike-globalfunkradio)
Awesome! good news that :)

**Comment by Holzhaus** [comment]
(by rryan)
Well, it's scheduled to 1.12.0 but we don't have anyone working on it. I may take a look.

**Comment by Holzhaus** [comment]
(by daschuer)
Probably related:
Bug #667651
Bug #726894
Bug #1116474
Bug #1269139

**Comment by Holzhaus** [comment]
(by rryan)
Any progress on this? What techniques do you plan to try? 

My suggestion is to write the secondary (non-master) samples to a FIFO when the master is processed and then to use a soundstretcher (e.g. rubberband) to slightly rate adjust to adjust for the clock drift. This is the more advanced way to do this -- there may be a simpler way? 

Also, we need criteria for when to abandon this for 1.12.0 as it should not delay the release. I propose a deadline of April 1 (beta1)?

**Comment by Holzhaus** [comment]
(by daschuer)
Yes, work is in progress: 
https://github.com/daschuer/mixxx/tree/two_soundcards
I plan to have a fist prototype after this weekend. 

I will use the Portaudio Read/Write I/O Method from the clock reference device callback. 
The Buffers are already in place.
IMHO a soundstretcher is not required, because we are talking only about a few samples. 


In my fist version I will just chop the buffer, or duplicate copy of the last sample.
Perforce over quality for Headphones.  

There are two pending problems:
* Latency sync -> I have reanimated Owns delay patch Bug #225966 
* Underflow:
  -  Bug #1198306 
What is the best way to catch up, without use extra CPU time?

**Comment by Holzhaus** [comment]
(by rryan)
> I will use the Portaudio Read/Write I/O Method from the clock reference device callback.

You mean you will call Pa_WriteStream from within the clock reference device callback? Pa_WriteStream blocks until the OS is ready to receive samples for that stream. I don't believe this will solve the problem -- worse I think it will cause xruns from the clock reference callback since it will block our existing callback.

Can you explain why that would solve the problem?

**Comment by Holzhaus** [comment]
(by rryan)
Even if you used Pa_GetStreamWriteAvailable to avoid blocking, this would not solve the fundamental problem of this bug report (as I understand it) which is that you have two clocks progressing at different rates. 

If the headphone output is being consumed slower than the master output, you will eventually have to queue up samples or cause corruption in the audio by dropping samples. If the clock is slower then the headphone output will never catch up. Similarly if the headphone clock is faster, you will always under-run.

The only solution I see is to 1) detect the clock skew and 2) time-stretch to adjust for it.

Unless I'm wrong about the cause of this bug (differing clocks) -- I don't see how what you mentioned solves the problem. What have you found from your investigations?

**Comment by Holzhaus** [comment]
(by rryan)
Also, note the portaudio documentation specifically says not to do this:

> @note With the exception of Pa_GetStreamCpuLoad() it is not permissible to call
> PortAudio API functions from within the stream callback.

It may not cause problems to make API calls for a different stream though.

**Comment by Holzhaus** [comment]
(by daschuer)
I do not like the idea to have more than one real time thread. 
The clock reference device produces the audio Samples. So it is a good idea to feed them to the second device just after we have them. This will still allow minimum latency. 

I have not verify it. But in case both cards are perfectly synced,  Pa_GetStreamWriteAvailable should always give the exact size of the defined audio buffer. In case this is smaller or bigger we can drop or add samples, or stretch the  buffer if it turns out to be notable.

If this does not work, I will go the way you mention and measure the distortion.

**Comment by Holzhaus** [comment]
(by rryan)
Regardless I don't think going against the PA documentation is a good idea. We shouldn't use Pa_WriteStream from a callback. Even if it doesn't crash on your host API, we won't know that it doesn't on all platforms and APIs.

**Comment by Holzhaus** [comment]
(by rryan)
I think first we need data on what the problem actually is. Is it clock skew or is it just a timing issue (the non-reference callback occurs before the clock reference is done producing this epoch's data).

I propose you accomplish this by writing to a FIFO when the master clock reference is processed and then when the non-reference callback occurs, read from the FIFO. That produces an easy measurement framework for measuring the clock skew -- the queue length - desired samples is the clock skew.

**Comment by Holzhaus** [comment]
(by rryan)
You can also use the event timeline functionality I added to produce an event log of the process calls from the soundcards. You could whip up a quick python script to analyze the frequencies of the callbacks to see if they are different or drift relative to each other.

**Comment by Holzhaus** [comment]
(by daschuer)
The functions are returning  paCanNotReadFromACallbackStream or paCanNotWriteToACallbackStream if called from a callback even from a different device.

**Comment by Holzhaus** [comment]
(by rryan)
I think that's what happens when a stream is opened in callback mode and you call Pa_WriteStream on it. For non-reference-clock streams you would have to open them in non-callback mode for that to work. 

But still, if the PA docs say don't do it, we shouldn't do it.

**Comment by Holzhaus** [comment]
(by rryan)
At least I noticed that when reading the ALSA hostapi.

**Comment by Holzhaus** [comment]
(by rryan)
I've been testing this on my hardware. Using a built in soundcard and my NI Audio 10 the problem seems to be at least concurrent callbacks and maybe skew. The callbacks occur at different rates -- every now and then two NI Audio callbacks happen for one built-in and vice-versa.

See attached timeline.

**Comment by Holzhaus** [comment]
(by rryan)
For a while I was confused why we don't see similar issues with input devices. Then I realized all input handlers write their contents to a FIFO immediately. I'm curious to see if we see underflows or overflows in e.g. microphone or VC inputs when using heterogenous soundcards.

**Comment by Holzhaus** [comment]
(by daschuer)
Since in and output is in the same callback there must be the same issue. But IMHO you can't "hear" the effect, because you can't hear inputs and the output is krackling anyway in the same moment. 

I will now try again to read/write to the second sound card from the fist sound card's callback. If this remains locked by port audio, I will introduce a new thread, synced with the callback thread. IMHO it should be the same from the PA point of view, using just a different thread ID, so it cant be a real issue. Each sound card should live in its own memory space. 
Once I have some experience, I will discuss the possible solutions on the Portaudio mailing list.

**Comment by Holzhaus** [comment]
(by daschuer)
Here is the solution for our problem: 
http://quicktoots.linux-audio.com/toots/el-cheapo/
But be warned: soldering is involved. ;-)

**Comment by Holzhaus** [comment]
(by rryan)
@daschuer -- do you see what I mean though... we will have no way to know that your use of Pa_WriteStream within a callback doesn't cause issues on some platform / API combo. Asking PA developers what the risks are is a good idea.

**Comment by Holzhaus** [comment]
(by rryan)
Daniel -- I'm sorry to step in on your work on this bug but I think you're on the wrong track.

We already have all the tools we need to solve this bug within Mixxx. You don't need a new thread -- we already have an execution context for the other soundcard (the callback). 

Using a FIFO approach allows us to do the work needed for timing adjustments (time scaling) in the callback for the card that needs it. This has the benefit of avoiding unnecessary work in the clock reference callback. 

I have prepared a small patch that implements what I'm suggesting. Surprisingly, it actually fixes the issue for me on my built-in internal soundcard and my Audio 10 -- no crackling at all can be heard in my headphones where without the patch it is unusable due to the crackling. I guess I am not affected by clock skew (at least not a the timescales I am testing). 

If you are using a single soundcard, performance is not affected at all. Output buffer composition uses the AudioOutput buffer directly if the device is the clock reference device.

https://github.com/mixxxdj/mixxx/pull/199

**Comment by Holzhaus** [comment]
(by daschuer)
Here is the PA post http://music.columbia.edu/pipermail/portaudio/2014-March/015890.html

**Comment by Holzhaus** [comment]
(by daschuer)
Answer from Ross: 


> Can I assume for now, that once the timing issues are solved there are
> no objections to call read/write from a callback thread of a different
> device (PaStream)?


No. You can not, in general, safely call Pa_ReadStream()/Pa_WriteStream() from within another PA callback.

In particuar Pa_ReadStream/Pa_WriteStream may block if their internal buffers are full. There is also no guarantee that they are real-time safe calls.

My recommendation would be to use callback mode for both streams and build your own buffering between them (maybe using pa_ringbuffer)

**Comment by Holzhaus** [comment]
(by rryan)
Ok! We finally have a band-aid fix that should help people who were seeing concurrent callbacks. In the future we can add time-stretching or shrinking to adjust for clock skew.

**Comment by Holzhaus** [comment]
(by rryan)
Filed a bug for the future solution: Bug #1293800. Marking this one fixed as it will help users who are facing concurrent callbacks but minor skew.
